### PR TITLE
Implement Google Sign In

### LINKED ISSUE #4: Full login/signup functionality
Implement Firebase Auth in the same way as supported on the old Dante Android App. Users should be able to login with Google, Email and anonymously. Furthermore, anonymous users should be able to upgrade to an email account later on.

One special requirement is to use the `Google Sign-In` dependency for Google signup, as these credentials are later on required to access the Google Drive backups

- [x] Login with Google 
- [x] Login anonymously 
- [x] Login with Email
- [x] Upgrade from anonymous user to email user

### PR DESCRIPTION
**Pull request by lockieRichter**
Having a go at implementing the google sign in, but I'm new to the BLoC pattern so not sure I've done this correctly. Looking for feedback on this.

### CODE REVIEW

File: lib/src/bloc/auth/login_bloc.dart [hunk 1]
```diff
@@ -9,11 +9,11 @@ class LoginBloc {
     return (await _repository.getAccount()) != null;
   }
 
-  void loginWithGoogle() {
+  void loginWithGoogle({required Function onSuccess}) {
     _repository.loginWithGoogle().then(
       (value) {
         print('Logged in with Google!');
-        // TODO Update UI
+        onSuccess();
```
Comment (lockieRichter) [inline lib/src/bloc/auth/login_bloc.dart]: Passing in the navigation as a callback here doesn't feel right as we would have to pass in the same callback for the other sign in methods. Unsure of what the best way to do it is though...

Comment (shockbytes) [reply lib/src/bloc/auth/login_bloc.dart]: Would make sense, but I would rather decouple this fully with a stream/Rx Subject that relays this to the UI.

Something like this:

```Dart
  
  final PublishSubject<LoginEvent> _loginSubject = PublishSubject();
  Stream<LoginEvent> get loginEvents => _loginSubject.stream; 

  void loginWithGoogle() {
    _repository.loginWithGoogle().then(
      (value) => _loginSubject.add(GoogleLoginEvent())
    );
  }
```

This has in my opinion 2 benefits:

1.) We can listen to UI updates anywhere in the widget and do not need to handle it directly.
2.) We don't really need to rely on the widget's state but rather use stateless widgets most of the time.

In this case we would make `LoginPage` a `StatefulWidget` and register a stream listener in `initState`. If there comes any login event (Google, Mail, Anonymous), we can then navigate to the main page.

What do you think?

File: lib/src/bloc/auth/logout_bloc.dart [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+class LogoutBloc {
+  final AuthenticationRepository _repository;
+
+  LogoutBloc(this._repository);
+
+  void logout() {
+    _repository.logout().then(
+      (value) {
+        print('Logged out!');
```
Comment (shockbytes) [inline lib/src/bloc/auth/logout_bloc.dart]: Here we can do the same thing and expose the result as a stream. We can listen to logout events throughout the app by just listening to the stream then.

**Review by shockbytes — COMMENTED** [review]
Looks good, mainly one open discussion how to listen to callbacks within the bloc.

File: lib/src/bloc/auth/login_bloc.dart (hunk 1 above)
Comment (lockieRichter) [reply lib/src/bloc/auth/login_bloc.dart]: Yes I see how you do it now, let me have a go.

File: lib/src/bloc/auth/logout_bloc.dart (hunk 2 above)
Comment (lockieRichter) [reply lib/src/bloc/auth/logout_bloc.dart]: Done.

**Comment by lockieRichter** [comment]
@shockbytes addressed your comments, take a look when you can

File: lib/src/bloc/auth/login_bloc.dart [hunk 3]
```diff
@@ -1,19 +1,24 @@
...
     return (await _repository.getAccount()) != null;
   }
 
   void loginWithGoogle() {
     _repository.loginWithGoogle().then(
-      (value) {
-        print('Logged in with Google!');
-        // TODO Update UI
+      (_) {
+        _loginSubject.add(LoginEvent.googleLogin);
```
Comment (shockbytes) [inline lib/src/bloc/auth/login_bloc.dart:21]: I think we should also handle login errors here by forwarding a `googleLoginError` event, what do you think?

Comment (lockieRichter) [reply lib/src/bloc/auth/login_bloc.dart:21]: Good idea!

**Review by shockbytes — APPROVED** [review]
Nicely done 🎉

**Pull request merged**